### PR TITLE
Sridha2412 1

### PR DESCRIPTION
**Pull request by Sridha2412**

### CODE REVIEW

File: sridha2412/assignment1/string_anagrams.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+# assumption: only using the 26 alphabets of the english language
+# method to check if two strings are anagrams
+# method takes two strings and a boolean value. If true, check case sensitive anagrams else ignore case
+def isAnagram(str1, str2, caseSensitive):
+    # if ignoring case, convert both strings to lower case
+    if not caseSensitive:
+        str1 = str1.lower()
+        str2 = str2.lower()
+    # sort and compare the strings
+    return sorted(str1) == sorted(str2)
```
Comment (jdb) [inline sridha2412/assignment1/string_anagrams.py]: I see a duplication of the logic: the same processing is done on str1 and str2, and then below on each of the words of each of the sentence. Do you think you can propose a solution to encapsulate this logic to be able to reuse it?

File: sridha2412/assignment1/string_anagrams.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+# assumption: sentence can only include alphabets and spaces
+# method to check if two sentences are anagrams
+# method takes two sentences and a boolean value. If true, check case sensitive anagrams else ignore case
+def isSentenceAnagram(str1, str2, caseSensitive):
```
Comment (jdb) [inline sridha2412/assignment1/string_anagrams.py]: I think a required condition for two sentences to be anagrams is for each of the word in sentence1 to have an anagram word in sentence2. Do you think you can re-use isAnagram() in IsSentenceAnagram?

File: sridha2412/assignment1/string_anagrams.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    # sort and compare the strings
+    return sorted(str1) == sorted(str2)
+
+str1 = input("Enter first string: ")
+str2 = input("Enter second string: ")
+caseSensitive = input("Enter Y if case sensitive else N: ")
+# default value is False
+caseSensitive = True if caseSensitive == 'Y' else False
+
+# for words
```
Comment (jdb) [inline sridha2412/assignment1/string_anagrams.py]: Please remove the "dead" code

File: sridha2412/assignment1/string_anagrams.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+# python program to determine if two strings are anagrams of each other
+
+# assumption: only using the 26 alphabets of the english language
+# method to check if two strings are anagrams
+# method takes two strings and a boolean value. If true, check case sensitive anagrams else ignore case
+def isAnagram(str1, str2, caseSensitive):
```
Comment (jdb) [inline sridha2412/assignment1/string_anagrams.py]: Please check out https://google.github.io/styleguide/pyguide.html#Naming

Python usually do not use a lot the "camel case" convention but instead uses lower case and underscore.

File: sridha2412/assignment1/string_anagrams.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+str1 = input("Enter first string: ")
+str2 = input("Enter second string: ")
+caseSensitive = input("Enter Y if case sensitive else N: ")
+# default value is False
+caseSensitive = True if caseSensitive == 'Y' else False
+
+# for words
+# if isAnagram(str1, str2, caseSensitive):
+# for sentences
+if isSentenceAnagram(str1, str2, caseSensitive):
```
Comment (jdb) [inline sridha2412/assignment1/string_anagrams.py]: Think about using the ternary operator, you would get less duplication
is_it_or_not = '' if isSentenceAnagram(str1, str2, caseSensitive) else 'NOT '
print(...)

File: sridha2412/assignment1/string_anagrams.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    print("\""+str1+"\" and \""+str2+"\" are anagrams")
```
Comment (jdb) [inline sridha2412/assignment1/string_anagrams.py]: No need to the slash: you can alternate the double and single quote: this is quite convenient.

Ex:
print("I'm embedding a single quote without slashes.")

print('What "they" want us to believe is that "we" "landed" on "the moon"')
print('Wake up, ' + "Sheeple !")

**Review by jdb — CHANGES_REQUESTED** [review]
Very cool, my first CodeU pull request, good job you :)

File: sridha2412/assignment1/string_anagrams.py (hunk 2 above)
Comment (Sridha2412) [reply sridha2412/assignment1/string_anagrams.py]: I have made all necessary changes. However, I am not sure sentence_anagram(..) is very efficient.

File: sridha2412/assignment1/string_anagrams.py [hunk 7]
```diff
@@ -14,35 +14,31 @@ def isAnagram(str1, str2, caseSensitive):
...
-    for i, word in enumerate(str2):
-        str2[i] = sorted(list(word))
-
-    # sort and compare the strings
-    return sorted(str1) == sorted(str2)
+    # check both have equal number of words
+    if len(str1) != len(str2):
+        return False
+    # iterate over both lists
+    for i in range(0, len(str1)):
```
Comment (jdb) [inline sridha2412/assignment1/string_anagrams.py]: 2 notes: 
- range(0, 10) is equivalent to range(10). When the first element is not set, then its value defaults to zero.
- this is a problematic change: the previous algorithm was O(n * ln (n)) but now it is O(n * m) where n is the number of words in the first sentences and is the number of words in the second sentence.

I suggest you create a function that takes a sentence and returns a structure whose equality method test for anagrams. This function can take the sentence as an input and return a list of occurences count dictionaries. Do not forget to sort the words of the sentence before creating the list.

I will send a solution of anagram today.

**Review by jdb — CHANGES_REQUESTED** [review]
Ah, there is an introduction of O(n^2) algorithm over the previous version that was O(n * ln (n)). This is a regression :)

**Review by jdb — APPROVED** [review]

**Pull request merged**